### PR TITLE
feat(traceloop-sdk): add sampling_rate parameter to control trace volume

### LINKED ISSUE #2109: 🚀 Feature: Support for trace sampling
### Which component is this feature for?

Traceloop SDK

### 🔖 Feature description

Traces should have support for setting sampling rate.

### 🎤 Why is this feature needed ?

Sampling rate helps control the trace volume and improves performance for very high throughput or latency sensitive applications. Also, limiting trace volume helps control cost.

### ✌️ How do you aim to achieve this?

Similar to https://langfuse.com/docs/tracing-features/sampling or via configuration

### 🔄️ Additional Information

_No response_

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### Are you willing to submit PR?

None

### PR DESCRIPTION
**Pull request by Nikhil172913832**
Add a simplified sampling_rate parameter to Traceloop.init() for easy trace sampling configuration without manually creating sampler objects.

Closes #2109

- [x] I have added tests that cover my changes.
- [ ] If adding a new instrumentation or changing an existing one, I've added screenshots from some observability platform showing the change.
- [x] PR name follows conventional commits format: `feat(instrumentation): ...` or `fix(instrumentation): ...`.
- [ ] (If applicable) I have updated the documentation accordingly.

----

> [!IMPORTANT]
> Add `sampling_rate` parameter to `Traceloop.init()` for trace sampling control, with validation and conflict checks, and corresponding tests.
> 
>   - **New Features**:
>     - Added `sampling_rate` parameter to `Traceloop.init()` in `__init__.py` to control trace volume (0.0–1.0).
>     - Raises error if both `sampling_rate` and `sampler` are provided.
>     - Validates `sampling_rate` is between 0.0 and 1.0.
>   - **Tests**:
>     - Added `test_sampling_rate.py` to test initialization with `sampling_rate`, invalid values, conflict with `sampler`, and edge cases (0.0 and 1.0).
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=traceloop%2Fopenllmetry&utm_source=github&utm_medium=referral)<sup> for fdb5d632750c35a01e2d18bac9a6f6b150452597. You can [customize](https://app.ellipsis.dev/traceloop/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

----

## Summary by CodeRabbit

* **New Features**
  * Added a sampling-rate option to SDK initialization to control trace volume (0.0–1.0). Supplying a sampling rate will configure an internal ratio-based sampler; providing both a sampling rate and a custom sampler is disallowed and raises an error.
* **Tests**
  * Added tests covering initialization with sampling rate, invalid values, mutual-exclusivity with a custom sampler, and edge cases (0.0 and 1.0).

### CODE REVIEW

**Comment by Nikhil172913832** [comment]
Hi @nirga, just wanted to check if you could take a look at this PR when you get a chance.

File: packages/traceloop-sdk/traceloop/sdk/__init__.py [hunk 1]
```diff
@@ -136,6 +137,19 @@ def init(
...
+                "Cannot specify both 'sampler' and 'sampling_rate'. "
+                "Please use only one of these parameters."
+            )
+
+        if sampling_rate is not None:
+            if not 0.0 <= sampling_rate <= 1.0:
+                raise ValueError(
+                    f"sampling_rate must be between 0.0 and 1.0, got {sampling_rate}"
+                )
+            sampler = TraceIdRatioBased(sampling_rate)
```
Comment (nirga) [inline packages/traceloop-sdk/traceloop/sdk/__init__.py:143]: this is a no-op unless you also set it in the tracer provider:
https://opentelemetry-python.readthedocs.io/en/latest/sdk/trace.sampling.html

File: packages/traceloop-sdk/tests/test_sampling_rate.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+            exporter=exporter,
+            disable_batch=True
+        )
+
+        assert client is None
+        assert hasattr(TracerWrapper, "instance")
+        assert TracerWrapper.instance is not None
+        
+        # Verify that a tracer provider was created with a sampler
+        tracer_provider = TracerWrapper.instance._TracerWrapper__tracer_provider
```
Comment (nirga) [inline packages/traceloop-sdk/tests/test_sampling_rate.py]: Don't use these low-quality autogenerated test - they don't really test anything. You should run a real-world test - like the other ones we have here

**Comment by Nikhil172913832** [comment]
Thanks for the review! I’ll update the PR with the suggested changes

**Comment by Nikhil172913832** [comment]
@nirga I might be missing something, but from my review it seems the sampling_rate does reach the TracerProvider.
Here’s the flow I traced:
User initializes with Traceloop.init(app_name="my_app", sampling_rate=0.5)
At line 150 in packages/traceloop-sdk/traceloop/sdk/__init__.py, a TraceIdRatioBased sampler is created
At line 161, that sampler is passed to the TraceWrapper
At line 90 in packages/traceloop-sdk/traceloop/sdk/tracing/tracing.py, it’s passed to init_tracer_provider
Finally, at line 417 in the same file, the sampler is passed to the TracerProvider

**Comment by nirga** [comment]
@Nikhil172913832 I think you forgot to push packages/traceloop-sdk/traceloop/sdk/tracing/tracing.py

**Comment by Nikhil172913832** [comment]
@nirga The lines I am referring to in packages/traceloop-sdk/traceloop/sdk/tracing/tracing.py already exist in the main branch — I haven’t made any edits to them in this PR.

**Comment by Nikhil172913832** [comment]
Regarding the tests, I took inspiration from test_sample_initialization to improve coverage and structure. Let me know if there’s anything else I can refine or expand.
```python
"""Tests for sampling_rate parameter functionality."""

import pytest
from opentelemetry.sdk.trace.sampling import TraceIdRatioBased
from opentelemetry.sdk.trace.export.in_memory_span_exporter import InMemorySpanExporter
from opentelemetry import trace
from traceloop.sdk import Traceloop
from traceloop.sdk.tracing.tracing import TracerWrapper


def test_sampling_rate_validation():
    """Test that sampling_rate validates input range."""
    exporter = InMemorySpanExporter()

    with pytest.raises(ValueError, match="sampling_rate must be between 0.0 and 1.0"):
        Traceloop.init(
            app_name="test-invalid-high",
            sampling_rate=1.5,
            exporter=exporter,
            disable_batch=True
        )

    with pytest.raises(ValueError, match="sampling_rate must be between 0.0 and 1.0"):
        Traceloop.init(
            app_name="test-invalid-low",
            sampling_rate=-0.1,
            exporter=exporter,
            disable_batch=True
        )


def test_sampling_rate_and_sampler_conflict():
    """Test that providing both sampling_rate and sampler raises ValueError."""
    exporter = InMemorySpanExporter()
    sampler = TraceIdRatioBased(0.5)

    with pytest.raises(ValueError, match="Cannot specify both 'sampler' and 'sampling_rate'"):
        Traceloop.init(
            app_name="test-conflict",
            sampling_rate=0.5,
            sampler=sampler,
            exporter=exporter,
            disable_batch=True
        )


def test_sampling_rate_creates_spans():
    """Test that sampling_rate parameter works correctly."""
    if hasattr(TracerWrapper, "instance"):
        _trace_wrapper_instance = TracerWrapper.instance
        del TracerWrapper.instance

    try:
        exporter = InMemorySpanExporter()
        Traceloop.init(
            app_name="test-sampling",
            sampling_rate=1.0,
            exporter=exporter,
            disable_batch=True
        )

        tracer = trace.get_tracer(__name__)
        for i in range(5):
            with tracer.start_as_current_span(f"test_span_{i}"):
                pass

        spans = exporter.get_finished_spans()
        test_spans = [s for s in spans if s.name.startswith("test_span_")]
        assert len(test_spans) == 5

    finally:
        exporter.clear()
        if '_trace_wrapper_instance' in locals():
            TracerWrapper.instance = _trace_wrapper_instance
'''

**Comment by nirga** [comment]
@Nikhil172913832 they're not here in the PR

**Comment by Nikhil172913832** [comment]
@nirga I have not yet committed the test. I wanted to have your opinion on their coverage and quality. I have tested them locally and they do pass.

**Comment by nirga** [comment]
@Nikhil172913832 i meant that the code itself is not complete. The tracing.py file is not here

**Comment by Nikhil172913832** [comment]
@nirga , I haven’t modified packages/traceloop-sdk/traceloop/sdk/tracing/tracing.py — the lines I referred to already exist in the main branch. My PR only adds the sampling_rate parameter and related logic in __init__.py along with the new tests.
I mentioned tracing.py just to show where the sampler is passed through, not as part of the changes here.

**Comment by Nikhil172913832** [comment]
Hi @nirga, I have updated the tests and cross verified that the sampling rate is being passed to the tracer provider as per your request. Let me know if any other changes are required.
